File: Labs/Lab 7/lab7.java
public int findTheCity(int n, int[][] edges, int distanceThreshold) {
    int[][] distance = new int[n][n];
    for (int i = 0; i < n; i++) {
        Arrays.fill(distance[i], Integer.MAX_VALUE);
        distance[i][i] = 0;
    }

    for (int[] edge : edges) {
        int from = edge[0];
        int to = edge[1];
        int weight = edge[2];
        distance[from][to] = weight;
        distance[to][from] = weight;
    }

    for (int k = 0; k < n; k++) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (distance[i][k] != Integer.MAX_VALUE && distance[k][j] != Integer.MAX_VALUE) {
                    distance[i][j] = Math.min(distance[i][j], distance[i][k] + distance[k][j]);
                }
            }
        }
    }

    int minCity = -1;
    int minReachable = n;
    
    for (int i = 0; i < n; i++) {
        int reachable = 0;
        for (int j = 0; j < n; j++) {
            if (i != j && distance[i][j] <= distanceThreshold) {
                reachable++;
            }
        }
        if (reachable <= minReachable) {
            minReachable = reachable;
            minCity = i;
        }
    }
    
    return minCity;
}

File: Labs/Lab 7/lab7.py
def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
    # Initialize the distance matrix with max value and 0 for diagonal elements
    distance = [[float('inf')] * n for _ in range(n)]
    for i in range(n):
        distance[i][i] = 0

    # Populate the distance matrix with edge weights
    for from_city, to_city, weight in edges:
        distance[from_city][to_city] = weight
        distance[to_city][from_city] = weight

    # Floyd-Warshall algorithm to find the shortest paths between all pairs of cities
    for k in range(n):
        for i in range(n):
            for j in range(n):
                distance[i][j] = min(distance[i][j], distance[i][k] + distance[k][j])

    min_city = -1
    min_reachable = n

    # Iterate through all the cities and count reachable cities
    for i in range(n):
        reachable = sum(1 for j in range(n) if i != j and distance[i][j] <= distanceThreshold)
        if reachable <= min_reachable:
            min_reachable = reachable
            min_city = i

    return min_city
